### DISCUSSION TITLE
streams-generator-a2dp and stream mixing

### DISCUSSION THREAD

**Discussion by Ince1956** [pschatzmann/arduino-audio-tools#544] [Q&A] [answered]
I have tried the streams-generator-a2dp example and although it connects to the Bluetooth speaker I don't get any sound.   the debug output shows
[D] AudioA2DP.h : 149 - waiting for connection
[D] AudioA2DP.h : 295 - static void audio_tools::A2DPStream::a2dpStateCallback(esp_a2d_connection_state_t, void*)
[W] AudioA2DP.h : 300 - ==> state: Connected
[D] AudioA2DP.h : 149 - waiting for connection
[D] AudioA2DP.h : 149 - waiting for connection
[D] AudioA2DP.h : 149 - waiting for connection
[D] AudioA2DP.h : 149 - waiting for connection
[D] AudioA2DP.h : 149 - waiting for connection
[D] AudioA2DP.h : 149 - waiting for connection
[D] AudioA2DP.h : 149 - waiting for connection

However I can get sound with BluetoothA2DPSource using data from OneChannelSoundData

My aim is to have a number of sound samples in memory  send them to the Bluetooth speaker using the mixer.  I have tried streams-memory_pcm-mixer-audiokit and that works fine to the serial plotter.  What file format would be required in the MemoryStream to be able to send the output to the Bluetooth speaker.

**Comment by pschatzmann** [comment, marked as answer]
I recommend to use the Basic Callback API. You have much more control there! There are several examples in the A2DP project and in the Basic Examples of this project.

**Reply by Ince1956** [reply]
using base-adc-a2dp  I've tried replacing the call-back function with this  I still don't here anything from the speaker I have verified that it is being called with a Serial.println statement . Is there anything I am missing.  The sound file plays ok when I send with write_data.

SoundData *high_horn = new OneChannelSoundData((int16_t *)class_47_high_horn_raw, class_47_high_horn_raw_len / 2);
// callback used by A2DP to provide the sound data
int32_t get_sound_data(Frame* frames, int32_t frameCount) {
    uint8_t *data = (uint8_t*)frames;
    int frameSize = 2;
    size_t resultBytes =high_horn->get2ChannelData(frameCount*frameSize, frameSize,data);
    return resultBytes/frameSize;
}

**Reply by pschatzmann** [reply]
I suggest that you print the data to Serial so that you can double check what you send...
I am missing any logic which would keep track of the current position and send the remaining unprocessed samples and maybe would restart at the end.
